pkg/server/test/list_objects.go

```diff
@@ -51,6 +52,63 @@ func newListObjectsRequest(store, objectType, relation, user, modelID string, co
 func ListObjectsTest(t *testing.T, ds storage.OpenFGADatastore) {
 	ctx := context.Background()
 
+	t.Run("list_objects_with_concurrent_checks_and_contextual_tuples", func(t *testing.T) {
+		store := ulid.Make().String()
+
+		typedefs := parser.MustParse(`
+		type user
+
+		type doc
+		  relations
+		    define reader: [user] as self
+		    define blocked: [user] as self
+			define viewer as reader but not blocked
+		`)
+
+		model := &openfgapb.AuthorizationModel{
+			Id:              ulid.Make().String(),
+			SchemaVersion:   typesystem.SchemaVersion1_1,
+			TypeDefinitions: typedefs,
+		}
+		err := ds.WriteAuthorizationModel(ctx, store, model)
+		require.NoError(t, err)
+
+		typesys := typesystem.New(model)
+
+		listObjectsQuery := &commands.ListObjectsQuery{
+			Datastore:             ds,
+			Logger:                logger.NewNoopLogger(),
+			ListObjectsDeadline:   defaultListObjectsDeadline,
+			ListObjectsMaxResults: defaultListObjectsMaxResults,
+			ResolveNodeLimit:      defaultResolveNodeLimit,
+			CheckResolver:         graph.NewLocalChecker(storage.NewContextualTupleDatastore(ds), checkConcurrencyLimit),
+		}
+
+		request := &openfgapb.ListObjectsRequest{
+			StoreId:              store,
+			AuthorizationModelId: model.GetId(),
+			Type:                 "doc",
+			Relation:             "viewer",
+			User:                 "user:jon",
+			ContextualTuples: &openfgapb.ContextualTupleKeys{
+				TupleKeys: []*openfgapb.TupleKey{
+					tuple.NewTupleKey("doc:1", "reader", "user:jon"),
+					tuple.NewTupleKey("doc:2", "reader", "user:jon"),
+				},
+			},
+		}
+
+		ctx = typesystem.ContextWithTypesystem(ctx, typesys)
```

Comment:
a bit sad that we have to do this in the test instead of calling server's code to ensure this is setup properly :(

Reply:
Yep! That's why we need the functional tests for ListObjects. Our test entrypoint should be `server.ListObjects` instead of `commands.ListObjectsQuery`.